compiler/rustc_ast/src/token.rs

```diff
@@ -237,6 +237,15 @@ pub enum TokenKind {
     /// treat regular and interpolated lifetime identifiers in the same way.
     Lifetime(Symbol),
 
+    /// An embedded AST node, as produced by a macro. This only exists for
+    /// historical reasons. We'd like to get rid of it, for multiple reasons.
+    /// - It's conceptually very strange. Saying a token can contain an AST
+    ///   node is like saying, in natural language, that a word can contain a
+    ///   sentence.
```

Comment:
`proc_macro`-style tokens contain "sentences" (delimited groups) in them all the time, but for flattened `rustc_parse`-style tokens it's unusual, I guess.
When parser librarification was still in plans, one of the alternatives was to switch `rustc_parse` to the `proc_macro`-style token model.

The main reason to get rid of `Interpolated` tokens from my point of view is that we want to define everything happening during macro expansion in terms of token streams, so we have to treat the AST pieces as token streams at least in *some* cases, moreover the token stream representation is the source of truth for them.
So these AST pieces are supposed to be a parser caching mechanism at most, which is highly non-obvious when they are represented like this.

Reply:
You're talking about *token trees*, rather than tokens, right?

> The main reason to get rid of Interpolated tokens from my point of view ...

That's a good reason as well :)